src/findfile.c

```diff
@@ -506,107 +544,163 @@ vim_findfile_init(
 		}
 	    }
 	    else
-		ff_expand_buffer[len++] = *wc_part++;
+		ff_expand_buffer.string[ff_expand_buffer.length++] = *wc_part++;
 	}
-	ff_expand_buffer[len] = NUL;
-	search_ctx->ffsc_wc_path = vim_strsave(ff_expand_buffer);
+	ff_expand_buffer.string[ff_expand_buffer.length] = NUL;
 
-	if (search_ctx->ffsc_wc_path == NULL)
+	search_ctx->ffsc_wc_path.length = ff_expand_buffer.length;
+	search_ctx->ffsc_wc_path.string = vim_strnsave(ff_expand_buffer.string,
+	    search_ctx->ffsc_wc_path.length);
+	if (search_ctx->ffsc_wc_path.string == NULL)
 	    goto error_return;
     }
     else
-	search_ctx->ffsc_fix_path = vim_strsave(path);
+    {
+	search_ctx->ffsc_fix_path.length = STRLEN(path);
+	search_ctx->ffsc_fix_path.string = vim_strnsave(path,
+	    search_ctx->ffsc_fix_path.length);
+	if (search_ctx->ffsc_fix_path.string == NULL)
+	    goto error_return;
+    }
 
-    if (search_ctx->ffsc_start_dir == NULL)
+    if (search_ctx->ffsc_start_dir.string == NULL)
     {
 	// store the fix part as startdir.
 	// This is needed if the parameter path is fully qualified.
-	search_ctx->ffsc_start_dir = vim_strsave(search_ctx->ffsc_fix_path);
-	if (search_ctx->ffsc_start_dir == NULL)
+	search_ctx->ffsc_start_dir.length = search_ctx->ffsc_fix_path.length;
+	search_ctx->ffsc_start_dir.string = vim_strnsave(search_ctx->ffsc_fix_path.string,
+	    search_ctx->ffsc_start_dir.length);
+	if (search_ctx->ffsc_start_dir.string == NULL)
 	    goto error_return;
-	search_ctx->ffsc_fix_path[0] = NUL;
+	search_ctx->ffsc_fix_path.string[0] = NUL;
+	search_ctx->ffsc_fix_path.length = 0;
     }
 
+    if (search_ctx->ffsc_start_dir.length == 0
+	    && search_ctx->ffsc_start_dir.string != NULL)
+	search_ctx->ffsc_start_dir.length = STRLEN(search_ctx->ffsc_start_dir.string);
+    if (search_ctx->ffsc_fix_path.length == 0
+	    && search_ctx->ffsc_fix_path.string != NULL)
+	search_ctx->ffsc_fix_path.length = STRLEN(search_ctx->ffsc_fix_path.string);
+    if (search_ctx->ffsc_wc_path.length == 0
+	    && search_ctx->ffsc_wc_path.string != NULL)
+	search_ctx->ffsc_wc_path.length = STRLEN(search_ctx->ffsc_wc_path.string);
```

Comment:
Is this part ever needed? This may actually increase the number of `STRLEN` calls, and it seems that `length` is already properly set whenever these `string`s are modified.